src/mc-range-encoding.c

```diff
@@ -313,3 +313,303 @@ mc_getTypeInfoDouble (mc_getTypeInfoDouble_args_t args,
+   // Decimal128:
+   int32_t exp =
+      ((int32_t) mc_dec128_get_biased_exp (dec)) - MC_DEC128_EXPONENT_BIAS;
+   // We will scale up/down based on whether it is negative:
+   mlib_int128 e1 = mlib_int128_pow10 ((uint8_t) abs (exp));
+   if (exp < 0) {
+      ret = mlib_int128_div (ret, e1);
+   } else {
+      ret = mlib_int128_mul (ret, e1);
+   }
+
+   return ret;
+}
+
+bool
+mc_getTypeInfoDecimal128 (mc_getTypeInfoDecimal128_args_t args,
+                          mc_OSTType_Decimal128 *out,
+                          mongocrypt_status_t *status)
+{
+   /// Basic param checks
+   if (args.min.set != args.max.set || args.min.set != args.precision.set) {
+      CLIENT_ERR (
+         "min, max, and precision must all be set or must all be unset");
+      return false;
+   }
+
+   // We only accept normal numbers
+   if (mc_dec128_is_inf (args.value) || mc_dec128_is_nan (args.value)) {
+      CLIENT_ERR ("Infinity and Nan Decimal128 values are not supported.");
+      return false;
+   }
+
+   // Check boundary if a range is set
+   if (args.min.set) {
+      // [min,max] must be valid
+      if (mc_dec128_greater_equal (args.min.value, args.max.value)) {
+         CLIENT_ERR (
+            "The minimum value must be less than the maximum value, got "
+            "min: %s, max: %s",
+            mc_dec128_to_string (args.min.value).str,
+            mc_dec128_to_string (args.max.value).str);
+         return false;
+      }
+
+      // Value must be within [min,max)
+      if (mc_dec128_greater (args.value, args.max.value) ||
+          mc_dec128_less (args.value, args.min.value)) {
+         CLIENT_ERR ("Value must be greater than or equal to the minimum value "
+                     "and less than or equal to the maximum value, got "
+                     "min: %s, max: %s, value: %s",
+                     mc_dec128_to_string (args.min.value).str,
+                     mc_dec128_to_string (args.max.value).str,
+                     mc_dec128_to_string (args.value).str);
+         return false;
+      }
+   }
+
+   // Should we use precision mode?
+   //
+   // When we use precision mode, we try to represent as a decimal128 value that
+   // fits in [-2^127, 2^127] (i.e. is a valid int128)
+   //
+   // This check determines if we can represent any precision-truncated value as
+   // a 128-bit integer I.e. Is ((ub - lb) * 10^precision) < 128 bits.
+   //
+   // It is important that we determine whether a range and its precision would
+   // fit, regardless of the value to be encoded, because the encoding for
+   // precision-truncated-decimal128 is incompatible with the encoding of the
+   // full range.
+   bool use_precision_mode = false;
+   // The number of bits required to hold the result (used for precision mode)
+   uint8_t bits_range = 0;
+   if (args.precision.set) {
+      // Subnormal representations can support up to 5x10^-6182 as a number
+      if (args.precision.value < 0 || args.precision.value > 6182) {
```

Comment:
```suggestion
      if (args.precision.value > 6182) {
```

Fixes an observed warning in the C driver [patch build](https://spruce.mongodb.com/task/mongo_c_driver_clang34ubuntu_debug_compile_sasl_openssl_cse_patch_a8560d7488778e480b7c61a0f8931ddbcc310c7d_63c6c5185623437c083cb3c9_23_01_17_15_56_09/logs?execution=0) of libmongocrypt:

```
[2023/01/17 16:06:35.464] /data/mci/e5b1c47c15a7839221268eea1a500919/mongoc/libmongocrypt/src/mc-range-encoding.c:233:32: error: comparison of unsigned expression < 0 is always false [-Werror,-Wtautological-compare]
[2023/01/17 16:06:35.464]       if (args.precision.value < 0 || args.precision.value > 324) {
[2023/01/17 16:06:35.464]           ~~~~~~~~~~~~~~~~~~~~ ^ ~
[2023/01/17 16:06:35.466] /data/mci/e5b1c47c15a7839221268eea1a500919/mongoc/libmongocrypt/src/mc-range-encoding.c:412:32: error: comparison of unsigned expression < 0 is always false [-Werror,-Wtautological-compare]
[2023/01/17 16:06:35.466]       if (args.precision.value < 0 || args.precision.value > 6182) {
[2023/01/17 16:06:35.466]           ~~~~~~~~~~~~~~~~~~~~ ^ ~
[2023/01/17 16:06:35.471] 2 errors generated.
```

Can you also update line 233?

Reply:
Fixed